### ISSUE TITLE
fuzzy searching slow in big lists

### ISSUE THREAD

**Issue by alexanderjeurissen** [shoumodip/ido.nvim#8]
### Context
When trying to use `Ido` for filtering very big lists Ido slows down significantly. Potential culprit is filtering of the  `items` that ido receives. 

### Minimal config to reproduce

```lua
require "ido"

local RG_CMD = 'rg --files --ignore --smart-case --hidden --follow --no-messages --ignore-file ~/.gitignore'

function ido_ripgrep_example()
  return ido_complete {
    items = fn.systemlist(RG_CMD),
    on_enter = function(s) print(s) end
  }
end
```

When executing the above in a big git repository ido becomes almost unusable

### Potential causes / solutions

There are two things that currently slow down the filtering:

#### 1. We currently loop over the list of `items` twice:

Currently in order to determine both the `ido_matched_items` and `ido_true_matched_items` we loop over the items array twice. With big lists this can get very costly:

```lua
  ido_true_matched_items = table.filter(ido_match_list,
  function(v)
    if not ido_case_sensitive then
      v = v:lower()
    end

    if v:match('^' .. true_ido_pattern_text) then
      return true
    end
  end
  )

  ido_matched_items = table.filter(ido_match_list,
  function(v)
    if not ido_case_sensitive then
      v = v:lower()
    end

    if v:match(ido_pattern_text) and not v:match('^' .. true_ido_pattern_text) then
      return true
    end
  end
  )
```

We can reduce the execution time significantly by looping over the collection once and adding to both collections within the single loop.

#### 2. Filtering in LUA that is JIT compiled is relatively slow compared to pre-compiled binaries

Filtering in Lua that is JIT compiled is relatively slow compared to pre-compiled binaries.

I know other fuzzy searchers suffer the same issue (notably nvim-telescope https://github.com/nvim-telescope/telescope.nvim/issues/392) 

Utilizing a native compiled binary might help in this case. That could be fzf or something along the lines of 

https://github.com/romgrk/fzy-lua-native

However this is quite a major lift as it would require rewriting the prefix method as well.

### Suggestion

This would be a great opportunity to extend the api of ido and allow the user to provide a custom filter function. That way we can keep the default filter behavior and allow users to define their own filter functions either in their configuration or as separate plugins/libraries.

**Comment by shoumodip** [comment]
I have been experimenting with the LuaJIT bindings to the fzy algorithm with very promising results. I think I will make that the default algorithm in Ido.

- It yields far better results than the current implementation, with proper scoring
- It really simplifies the suggestion feature (prefix)
- I will be able to implement a character highlight feature using this.

I don't think an option is needed for this, fzy is the clear winner in all situations. It will simplify a lot of the code, make the searching insanely fast, and score better. Then again, this is not the final option for now. Let me know what you think about this. Also I'm really sorry but I won't be able to make any changes in like a month. I'm currently busy with exams.

Also I really appreciate the fact that you took your time to come with a solution to improve Ido. Of course whether I merge it in depends on what the final choice of the algorithm is. Again, let me know whether fzy should be the final choice or should it be an option. I personally am more inclined towards the former, as it will simplify a *lot* of code, but then again, I am really willing to take your choice into account.

**Comment by alexanderjeurissen** [comment]
Hi @shoumodip

Thanks for the kind words and reply.

You got me, that was my intent behind this change yes. Making FZY the default is the obvious choice.
Tbh the reason I opted for making it configurable in the PR is because I anticipated it to be the least controversial and quickest way to remain backwards compatible and to expose the necessary API for me to write my own FZY filter function 😅.

Given that you are open for going the extra mile and making `FZY` the default algorithm I take it you would be open for a PR that implements that ? I might have some time to create such a PR while you focus on your exams ;-)
